### PR TITLE
ARROW-1163: [Java] Java client support for plasma

### PR DESCRIPTION
**Pull request by salah-man**
--------------------------
This commit includes the support of java client for plasma, which is part of the java worker support of Ray.
In addition to some minor changes in build system, it consists of the following modules:
 - java/plasma: java client support for plasma
 - cpp/src/plasma/lib/java: JNI support for plasma client
 - java/plasma/Readme.md: Readme.md for help document
 - java/plasma/test.sh: test.sh for test

### CODE REVIEW

**Comment by robertnishihara** [comment]
Related to ray-project/ray#2095.

**Comment by jacques-n** [comment]
I think should integrate better into the other Arrow Java APIs. Right now, the apis here don't interact with Java ValueVector nor ArrowBufs, the key components of the Java API. We should make sure to integrate at that level so that our pieces work well together.

**Comment by jacques-n** [comment]
Also, would be good to have a jira ticket tracking this discussion/approach.

**Review by pcmoritz — COMMENTED** [review]
Thanks for the contribution! This looks really great :)

Before it can be merged, we should have a few things:
- Fix all the linting/test errors on Travis (e.g. org_apache_arrow_plasma_PlasmaClientJNI.cc needs a license header, <jni.h> is not found in the C++ build, so we need to install it, there is probably some linting error to be fixed once the license is there
- Add some tests to make sure the Java PlasmaClient is working
- Provide instructions on how to set it up and compile it for developers

I think the things @jacques-n brings up can be addressed in a follow-up PR, let's make sure the low-level APIs are working well in this PR.

**Comment by salah-man** [comment]
Thank you for your time @pcmoritz. The comment from you really helpful.
1. I will fix all the linting/test errors on Travis.
2. I think it is really needed for the test case, and for the java client of plasma, maybe we need to start the standalone process of plasma store to do the test which i will take a moment to try. And do you have any suggestion here?
3. I will add the Readme.md under the package of plasma for the usage guide

**Comment by pcmoritz** [comment]
Thanks, that sounds good!

For 2, we can do the setup similar to how we do it in C++ (see https://github.com/apache/arrow/blob/master/cpp/src/plasma/test/client_tests.cc#L61) and start an instance of the store from the Java tests (and shut it down after the tests are done).

**Comment by jacques-n** [comment]
The public interface (which is fine in a separate patch), should be ValueVector & ArrowBuf based, not byte[] based. Before releasing this functionality we should be making sure that it works in the context of the rest of the Arrow Java code.

**Comment by robertnishihara** [comment]
In Python, we expose two different object store client APIs.
1. An API for creating/sealing/getting a writeable buffer (I think it's called `PlasmaBuffer`).
2. An API for storing/retrieving a Python object.

In Java, I think it makes sense to have analogs of these. The more important one for now is the first one, which could be implemented using Arrow buffers (but the stored data would be blobs and not necessarily need to be data serialized using the Arrow format). The second would use the Arrow format under the hood.

**Comment by salah-man** [comment]
Yes, i think it is make sense to hava analogs of these the same as python, and we could try to use the Arrow buffers to instead the byte[] in the follow-up PR, which could make sure that it works in the context of the rest of the Arrow Java code. Thank you for your time again, @pcmoritz @jacques-n  @robertnishihara

**Comment by salah-man** [comment]
As comment, we just fix the problem you have suggest, could you review them again? Thank you very much, @pcmoritz

File: java/plasma/test.sh
Comment (pcmoritz) [inline java/plasma/test.sh]: I had to do this to make it compile with Ubuntu 16.04, since the boost static libraries that ship with it are compiled without -fPIC.

File: cpp/src/plasma/CMakeLists.txt
Comment (pcmoritz) [inline cpp/src/plasma/CMakeLists.txt]: Why do we need the ` -Wl,-force_load,` part here?

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: Let's add this file to the C++ linting/clang-tidy

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: should this be a constexpr?

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: Let's do a reinterpret_cast here, we shouldn't have old style casts

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: ditto

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: This class needs to be integrated into this PR, right? Can we also add a test that this works?

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: same here

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: C style cast

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: yes, it will not be changed, and worked as constant

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: C style cast

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: Let's add a JIRA that we should figure out a way to propagate arrow::Status to java automatically

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: ok

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: C style cast

File: java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java
Comment (pcmoritz) [inline java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java]: can we add tests for "contains" and "hash" here?

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: yeah, the test case current add to the ci will test this file too, so the current ci is passed means this class is work as well.

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: ok, i will add a JIRL later

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (pcmoritz) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: Thanks! Also, naming it OBJECT_ID_SIZE would be more consistent with the rest of the code (but up to you, since it's private in this file)

**Review by pcmoritz — COMMENTED** [review]
Thanks a lot! I pushed some small changes and also made some comments :)

Let's also see if any Java arrow developers are interested in reviewing it.

File: cpp/src/plasma/CMakeLists.txt
Comment (salah-man) [reply cpp/src/plasma/CMakeLists.txt]: -Wl,-force_load for add runtime library path here

File: java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java
Comment (salah-man) [reply java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java]: If I remember correctly, 'contains' will use plasma_manager which is not exist here.

**Comment by salah-man** [comment]
Thank you very much @pcmoritz . I will fix the ploblem in the comment above soon.

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: Thanks, i could rename it by the way.

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: I know what you mean now, i will use the normal Exception(java.lang.Exception) instead, thanks.

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: I think this file has already in the C++ linting process now, because i just fix many linting errors which is feedback by the ci.

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: `constexpr` is preferable (since the compiler can optimize more)

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: c-style cast

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: c-style cast

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: c-style cast

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: Should these `NULL`s be `nullptr`? I'm not sure.

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: c-style cast

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: maybe just `return has_object`

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: remove unused code

If this is something that needs to be handled in the future, then add a comment with a TODO

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: casts

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: casts

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: casts

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: why is this commented out?

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: casts

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: casts

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: casts

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.h
Comment (robertnishihara) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.h]: @pcmoritz What should the comment style be in this file? For C++ files we've been doing `//` everywhere else, and using Doxygen style `///` for documentation.

However, should JNI files be handled differently?

Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.h]: Oh, it is generated by javah automatically

Comment (robertnishihara) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.h]: Ok, I see.

Comment (robertnishihara) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.h]: Does this PR include instructions for how to generate the files?

Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.h]: That is only the org_apache_arrow_plasma_PlasmaClientJNI.h will by generated by javah offline, and usually we do not need to add the  instructions for how to generate, because we have to write the code of  org_apache_arrow_plasma_PlasmaClientJNI.cc, and just check in the org_apache_arrow_plasma_PlasmaClientJNI.h as well.

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: it is

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: It is point here, maybe nullptr better, i will fix them.

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (salah-man) [reply cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: Yes, you are right, i am changing it to return has_object.

File: java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java
Comment (pcmoritz) [reply java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java]: Both contains and hash are just functions of the plasma_store, I think you meant the "Info" call

Comment (salah-man) [reply java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java]: Yes, you are right, i just test the contains and hash , both work well. I will add the test case soon

**Review by pcmoritz — APPROVED** [review]
+1 This looks good to me now. I'll merge unless there are more objections.

**Comment by pcmoritz** [comment]
@salah-man Are you or somebody from your team willing to take a stab at https://issues.apache.org/jira/browse/ARROW-2892? It is important and would be high impact I think (and you can probably do it much faster than me because I have limited experience with Java).

**Comment by wesm** [comment]
@salah-man can you let us know your JIRA id so we can assign this issue to you?

**Comment by raulchen** [comment]
hi, @wesm, this is Hao Chen from Ant Financial. Sorry for the delay, we have limited bandwidth currently.
Can you elaborate on what needs to be done? I can find someone from my team to work on this.

**Comment by jacques-n** [comment]
@salah-man, let's take this conversation to https://issues.apache.org/jira/browse/ARROW-2892

**Pull request closed without merging**
